Issue 2813: Ensure the correct Segment routing key range is updated with replacement range.

shrids opened the pull request:
**Change log description** 

    - Remove invalid assert in StreamSegments#verifyReplacementRange()
    - Ensure the correct segment range is updated within a StreamSegments#withReplacementRange().

**Purpose of the change**

Fixes #2813, and maybe #2826

**What the code does**

The following changes are performed in this pull request .

- The older code used to assert if the predecessors of the segment returned by the getSuccessor() call to the controller are in fact being replaced. This need not be always true as part of the segment range could be updated with the successor due to a previous invocation on an adjacent segment. This assertion has been removed.

- In-case a successor segment has a key range higher than the current segment then it could so happen that the newer writes are directed to the successor even before a segment sealed is encountered for the previous segment by the writer. The PR ensures that the newer writes to the segment happen only after the predecessor segments for the given routing key are sealed.

**How to verify it**

The existing and newly added tests should pass successfully.

tkaitchuck commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -71,13 +75,20 @@ public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replac
             } else {
                 for (SegmentWithRange replacement : replacements) {
-                    result.put(replacement.getHigh(), replacement.getSegment());
+                    /*
```
We also need to add a check and test for the case where (using the segment sequence from the unit test) 8 gets added before 4 is sealed so that we make sure that when 4 is added after the seal of 1 is encountered that 4 does not replace 8 in the map.

tkaitchuck commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -103,10 +116,7 @@ private void verifyReplacementRange(StreamSegmentsWithPredecessors replacementSe
-                                        "Inconsistant replaced replacement segments %s", replacementSegments);
-            Preconditions.checkArgument(replacementRanges.containsKey(lowerReplacedSegment.getValue().getSegmentId()),
-                                        "Inconsistant replaced replacement segments %s", replacementSegments);
+            // Note: replacedSegments can be already present in the current segment mapping in a scale down / merge scenario.
```
The check is more complex, but we should still make some assertion...

shiveshr replied:
definition: 
- `allocated range: range assigned to segment by controller`
- `included range: range included for a given segment in the writer's map`
Now assertions/invariants here should be: 
1. union of all `allocated ranges` (and same for `included ranges`) should  [0.0, 1.0). 
2. No segment should have an `included range` that exceeds its `allocated range`
3. replacement segment's `included range` should 
 - be a subset of replacement segment's `allocated range`
 - overlap with `exiting` segment's `allocated range`

tkaitchuck commented on client/src/test/java/io/pravega/client/stream/impl/StreamSegmentsTest.java:
```diff
@@ -124,6 +125,99 @@ public void testRangeReplacementMerge() {
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        // this should now return segment 8 and not segment 4
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.7));
```
At the end of the test 8 is actually in the map twice with two different ranges. That might not be an issue, but it seems like a lurking complexity that could become a bug in the future if the code changed without thinking about that.

shrids replied:
With the current code there will be two entries i.e
Key : 0.5 Value: `8.#epoch.4`
Key: 0.75 Value: `8.#epoch.4`

shiveshr replied:
then we should include only one range

tkaitchuck replied:
Yes, I worry about that when 8 is replaced. It is very easy to assume that it won't be in the list twice...

shrids replied:
updated the code to handle compaction. Will add a test case for the same.

tkaitchuck commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -89,6 +100,8 @@ public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replac
      * @param replacementSegments The StreamSegmentsWithPredecessors to verify
      */
     private void verifyReplacementRange(StreamSegmentsWithPredecessors replacementSegments) {
+        log.debug("Verification of replacement segments {} with the current segments {}",
```
Let's add a toString() on replacementSegments so this is a meaningful

shrids replied:
added it using @Lombok annotation.

arvindkandhare commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -71,13 +75,20 @@ public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replac
+                    /*
+                      In case of a segment merge the replacement segment's high key can be greater than or equal to the
+                      existing segment high key. In this scenario the existing segment's high key should be used.
+
```
Won't this leave a hole in the keyspace? Will help if you could add a sample example.

shrids replied:
Refactored the implementation based on the discussions in the PR; added junits samples for the same covering different scenarios.

shrids commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Ensure the correct Segment routing key range is updated with replacement range](https://trello.com/c/04sKTmsP/28-ensure-the-correct-segment-routing-key-range-is-updated-with-replacement-range)

tkaitchuck commented on client/src/test/java/io/pravega/client/stream/impl/StreamSegmentsTest.java:
```diff
@@ -93,38 +94,38 @@ public void testRangeReplacementSplit() {
-        assertEquals(20, counts[6]);
-    }
-    
+    //    @Test
```
uncomment

shrids replied:
done.

tkaitchuck commented on client/src/test/java/io/pravega/client/stream/impl/StreamSegmentsTest.java:
```diff
@@ -144,4 +145,127 @@ public void testSameRoutingKey() {
+    }
+
+    @Test
+    public void testRangeReplacementMultipleSegmentMerge() {
```
Let's also test the incremental version of this, where the seal is encountered for 2 then for 0 and then for 1. As this is awkward to deal with as the same segment ends up briefly with non-contiguous keyspace.

shrids replied:
added a test case.

shrids commented:
Update: System tests pass with this PR.

shiveshr commented:
@shrids https://github.com/pravega/pravega/pull/2814#pullrequestreview-143500110 this comment has not been addressed. Do you want to address it here or open a separate issue for it?

shrids commented:
@shiveshr I have addressed this comment and added a test case for the same ref: https://github.com/pravega/pravega/pull/2814/files#diff-8741a9414cdf5a6c5499aef1c74a7ad3R356

fpj commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -67,19 +71,28 @@ public Segment getSegmentForKey(double key) {
+        Segment lastSegmentValue = null;
+        for (Entry<Double, Segment> exitingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
+            final Segment exitingSegment = exitingEntry.getValue();
+            if (exitingSegment.equals(lastSegmentValue)) {
```
Why would we see the same segment twice?

tkaitchuck replied:
If the previous iteration of the loop already added this segment because that iteration of the loop was dealing with the second segment to seal following a merge.

fpj commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -67,19 +71,28 @@ public Segment getSegmentForKey(double key) {
-                result.put(exitingSegment.getKey(), exitingSegment.getValue());
+        Optional<List<SegmentWithRange>> replacements = Optional.ofNullable(replacedRanges.get(replacedSegment.getSegmentId()));
+        Segment lastSegmentValue = null;
+        for (Entry<Double, Segment> exitingEntry : segments.descendingMap().entrySet()) { // iterate from the highest key.
```
Is `exitingEntry` a typo? should it be `existingEntry`?

tkaitchuck replied:
yes

shrids replied:
corrected it.

fpj commented on client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java:
```diff
@@ -67,19 +71,28 @@ public Segment getSegmentForKey(double key) {
+            if (exitingSegment.equals(replacedSegment)) {
+                // segment needs to be replaced.
+                replacements.ifPresent(segmentWithRanges -> segmentWithRanges.forEach(segmentWithRange ->
+                        result.put(Math.min(segmentWithRange.getHigh(), exitingEntry.getKey()), segmentWithRange.getSegment())));
```
I'm not sure what kind of invariant we are trying to satisfy here. It looks like we are trying to guarantee that the we do not add an entry to the mapping with a high that is higher than the high of the existing entry, but why is that correct in the first place?

A related question is the following. It looks like we might end up with multiple entries in the map pointing to the same segment. This is the case so that we guarantee that the we receive segment sealed from each one of the existing segments before writing to a new segment that spans multiple existing ones. But, I'm wondering whether we need to consolidate the mapping once all segment sealed notifications are received.

tkaitchuck replied:
>I'm not sure what kind of invariant we are trying to satisfy here. It looks like we are trying to guarantee that the we do not add an entry to the mapping with a high that is higher than the high of the existing entry, but why is that correct in the first place?

it is possible it is only adding the segment for part of its range. For example if the new segment is a result of a merge and only one of the predecessors has encountered a seal.

>A related question is the following. It looks like we might end up with multiple entries in the map pointing to the same segment. This is the case so that we guarantee that the we receive segment sealed from each one of the existing segments before writing to a new segment that spans multiple existing ones. But, I'm wondering whether we need to consolidate the mapping once all segment sealed notifications are received.

yes, see above.

fpj replied:
I think the invariant is that we can only add to the result map ranges that fully overlap with the range of the replaced segment, in which case, I wonder whether it should be the min of `segmentWithRange` and `replacedSegment` instead.

In any case, I'd like to come to an agreement on what the precise invariant statement is and to see it expressed in the form of a comment in the code.

fpj replied:
Right, but this is very low level, I'm after the high-level property that we are trying to satisfy here and it should be expressed in terms of predecessor and successor ranges.

shrids replied:
```The successor segment(s)'s range should be limited to the replaced segment's range, there by ensuring that newer writes to the successor(s) happen only for the replaced segment's range.```

Does this look any better?
